keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+
+Currently there is no consistency when logging references to Kubenetes objects like Pods. Depending on a log just the name of a Pod, the name and namespace as separate fields or Pod name prefixed with a namespace and ‘/’ character is logged.
+This makes finding logs related to a given object much harder.
+To address this problem we want to implement helper functions for all standard object types which could be used to build values which when logged will provide a consistent identification of each of those objects.
+For instance for namespaced objects those methods should produce following value: “<namespace>/<name>”.
+
+Example:
+```go
+klog.InfoS("Pod status updated", "pod", metadata.Pod(pod), "status", "ready")
+```
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] “Pod status updated“ pod=“kube-system/kubedns“ status=“ready“
+```
+
+### Selecting most important logs
+
+As migration to new message structure will be done manually we will be focusing on logs that have the greatest impact for log querying and processing. We will be focusing on log messages Proposed plan of measurements:
+
+* Using log data from single run of kubernetes [gce-master-scale-performance](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance) tests
+* Based on logs generated on master node
+* Aggregate by components:
+  * kube-controller-manager
+  * kube-scheduler
+  * kube-apiserver
+  * kubelet
+* Only for log lines generated by klog (e.g. prefix `I0129 03:26:19.857798       1 garbagecollector.go:517]`)
+* Identified by code line that generated log (e.g. `garbagecollector.go:517`)
+
+We assume that exact details of taking measurements can be improved, but overall methodology is solid. With those criteria we concluded that covering 99.9% of logs on master node are generated by 22 log messages. Exact list of those logs is provided in the detail design section.
+
+#### JSON output format
+
+Introduction of new methods to klog library will make identification of different components of the log message much easier.
+With klog v2 we can take further advantage of this fact and add an option to produce structured logs in JSON format.
+
+Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```go
+klog.Infof("Updated pod %s status to ready", pod.name)
+```
+That would result in log
+
+```json
+{
+   "ts": 1580306777.04728,
+   "level": "info",
+   "msg": "Updated pod kubedns status to ready"
+}
+```
+
+And
+
+```go
+klog.InfoS("Pod status updated", "pod"", metadata.Pod(pod), "status"", "ready")
+```
+
+That would result in log
+```json
+{
+   "ts": 1580306777.04728,
+   "level": "info",
+   "msg": "Pod status updated",
+   "pod":{
```

Comment:
Should we really be this inconsistent from the "unstructured" format that would produce `pod="default/nginx-1"`? What's wrong with just having this non, nested `pod="nginx-1"` and `namespace="default"`?

Reply:
I guess that this is a bit a matter of opinion because at least for me the nested form is closer to the unstructured format as it keeps pod name and namespace together.
But apart from this there are some problems with non nested version like:
-  It breaks the one to one correspondence between InfoS method argument and fields in the JSON document.
- It makes problematic to have logs which  reference more than one namespace like for instance logging two pods each in a different namespace.
- It would be inconsistent with encoding of other structs passed as key values to InfoS method as those will be serialized as nested objects.